subprojects/core/src/main/java/org/gradle/api/internal/plugins/ImperativeOnlyPluginTarget.java

```diff
@@ -41,6 +49,28 @@ public void applyImperative(@Nullable String pluginId, Plugin<?> plugin) {
         // TODO validate that the plugin accepts this kind of argument
         Plugin<T> cast = uncheckedCast(plugin);
         cast.apply(target);
+
+        // TODO - this should probably be done with TypeMetadataWalker and some sort of annotation handler
+        if (target instanceof ExtensionAware) {
+            ExtensionContainer extensions = ((ExtensionAware) target).getExtensions();
+            Class<?> publicPluginType = GeneratedSubclasses.unpackType(plugin);
+            for (Method method : publicPluginType.getDeclaredMethods()) {
+                SoftwareType softwareType = method.getAnnotation(SoftwareType.class);
+                if (softwareType != null) {
+                    Object extension;
+                    try {
+                        extension = method.invoke(plugin);
+                    } catch (IllegalAccessException e) {
+                        throw new RuntimeException("Must be able to access @SoftwareType method", e);
+                    } catch (InvocationTargetException e) {
+                        throw new RuntimeException("Failed to create extension", e);
+                    }
+                    Class<?> returnType = softwareType.modelPublicType();
+                    extensions.add(returnType, softwareType.name(), Cast.uncheckedNonnullCast(extension));
```

Comment:
💭Consider a minor improvement: also report a user-friendly error message if the type of the actual return value cannot be cast to the statically declared public type.

Reply:
Yep, I'd like to capture this as a property metadata validation error.  Again, I plan to tackle this in the next round of changes.